subsys/bootloader/bl_boot/bl_boot.c

```diff
@@ -19,6 +19,55 @@
 #include <hal/nrf_gpio.h>
 #endif
 
+#if defined(CONFIG_SB_DISABLE_SELF_R_X)
+/* Disabling R_X has to be done while running from RAM for obvious reasons.
+ * Moreover as a last step before jumping to application it must work even after
+ * RAW has been cleared, therefore we are using custom RAM function relocator.
+ * Size of the relocated are isn't known but it doesn't matter as long as it contains at least
+ * entire aforementioned function.
+ */
+#define FUNCTION_BUFFER_LEN 128
+#define RRAMC_REGION_TO_LOCK_ADDR 0x5004b56c
+#define RRAMC_REGION_TO_LOCK_ADDR_H (RRAMC_REGION_TO_LOCK_ADDR >> 16)
+#define RRAMC_REGION_TO_LOCK_ADDR_L (RRAMC_REGION_TO_LOCK_ADDR & 0x0000fffful)
+typedef void (*ram_fn_ptr)(uint32_t);
+uint8_t ram_exec_buf[FUNCTION_BUFFER_LEN];
+
+static void bootconf_disable_r_x_and_jump(uint32_t vector)
+{
+	__asm__ volatile (
+		"   MOV r5, %0\n"
+		"   MOV r1, %1\n"
+		"   MOVT r1, %2\n"
+		"   LDR r0, [r1]\n"
+		/* Size of the region should be set at this point
+		 * by provisioning through BOOTCONF. If not set it according partition size.
+		 */
+		"   RORS r4, r0, #16\n"
+		"   CBNZ r4, clear_rwx\n"
+		"   MOVT r0, %3\n"
+		"clear_rwx:\n"
+		"   BFC r0, #0, #3\n"
+		"   STR r0, [r1]\n"
+		"   DSB\n"
+		"   BX r5\n"
+		:
+		: "r" (vector), "i" (RRAMC_REGION_TO_LOCK_ADDR_L),
+		  "i" (RRAMC_REGION_TO_LOCK_ADDR_H), "i" (CONFIG_PM_PARTITION_SIZE_B0_IMAGE / 1024)
+		: "r0", "r1", "r2", "r4", "r5", "memory"
+	);
+}
+
+static ram_fn_ptr relocate_bootconf_disable_r_x_and_jump(void)
```

Comment:
how this function preserves `r0` value?

Reply:
this function returns pointer to the callable one